lib/invoca/utils/guaranteed_utf8_string.rb

```diff
@@ -1,68 +1,107 @@
 # frozen_string_literal: true
 
-# This class expects to be initialized with a string and guarantees that the output of the to_string method is in UTF-8 format and fits in 3 bytes/char or less.
+# This class expects to be provides a normalize_string method that guarantees that the output of the to_string method is in UTF-8
+# format and fits in 3 bytes/char or less.
+#
+# [Deprecated] Equivalently, you can also create an instance of this class and call to_string or to_s on it.
 module Invoca
   module Utils
     class GuaranteedUTF8String
-      def initialize(string)
-        if string.is_a?(String) ||
-          (string.respond_to?(:to_s) &&
-           string.method(:to_s).owner != Kernel)  # the lame .to_s from Kernel just calls .inspect :(
-          @string = string.to_s
-        else
-          raise ArgumentError, "#{self.class} must be initialized with a string or an object with a non-Kernel .to_s method but instead was #{string.class} #{string.inspect}"
-        end
-      end
+      attr_reader :to_string
 
-      def to_string
-        @to_string ||= normalize_string(@string)
+      def initialize(string)
+        @to_string = self.class.normalize_string(string)
       end
 
-      alias_method :to_s, :to_string
+      alias to_s to_string
 
       private
 
       # chosen because this is a 1-byte ASCII character that is not used in any of the popular escaping systems: XML, HTML, HTTP URIs, HTTP Form Post, JSON
-      REPLACE_CHARACTER = '~' unless defined?(REPLACE_CHARACTER)
+      REPLACE_CHARACTER = '~'
 
-      def normalize_string(str)
-        str = @string.dup
-        str.force_encoding('UTF-8')
-        if !str.valid_encoding?
-          cp1252_to_utf_8(str)
+      class << self
+        def normalize_string(orig_string, normalize_utf16: true,
```

Comment:
Very minor, but do you want to put the second arg on its on line as well?

```suggestion
        def normalize_string(orig_string, 
                                           normalize_utf16: true,
```